### PR TITLE
pipeline: prevent trigger IPC timeout with xruns

### PR DESCRIPTION
**Pull request by ranj063**
START/RELEASE trigger IPC responses are now sent
from the pipeline task. When one of the connected
pipelines runs into an xrun even before the current
pipeline has a chance to run, the trigger IPC
will time out eventually. Fix this by returning
an error response. There is no need to trigger
the pipeline in this case as the pipeline
will be stopped by the host anyway.

Signed-off-by: Ranjani Sridharan <ranjani.sridharan@linux.intel.com>

### CODE REVIEW

**Review by lyakh — COMMENTED** [review]
Looks good in general, but let's think about one more case: #4724 which also moves STOP / PAUSE handling to the task. Would we also want to take this xrun path in those cases? Maybe it would be safer to use a `switch (p->trigger.cmd)` in this your patch to only handle START and RELEASE and see how STOP and PAUSE play out in the future? Or do we want to reply with an error in STOP in PAUSE cases too?

**Comment by ranj063** [comment]
> Looks good in general, but let's think about one more case: #4724 which also moves STOP / PAUSE handling to the task. Would we also want to take this xrun path in those cases? Maybe it would be safer to use a `switch (p->trigger.cmd)` in this your patch to only handle START and RELEASE and see how STOP and PAUSE play out in the future? Or do we want to reply with an error in STOP in PAUSE cases too?

good point @lyakh. I think we should reply in all cases.

**Comment by lyakh** [comment]
> > Looks good in general, but let's think about one more case: #4724 which also moves STOP / PAUSE handling to the task. Would we also want to take this xrun path in those cases? Maybe it would be safer to use a `switch (p->trigger.cmd)` in this your patch to only handle START and RELEASE and see how STOP and PAUSE play out in the future? Or do we want to reply with an error in STOP in PAUSE cases too?
> 
> good point @lyakh. I think we should reply in all cases.

but how would `xrun_bytes` become non-0 just in time for a `STOP` trigger? I guess it's pretty unlikely even if possible in theory? Or maybe it will definitely be caught at run-time?

**Comment by ranj063** [comment]
> but how would `xrun_bytes` become non-0 just in time for a `STOP` trigger? I guess it's pretty unlikely even if possible in theory? Or maybe it will definitely be caught at run-time?

I have no idea how this can happen but if it is never possible, why do we need the check at all? I think it is safer to send the error response in case of xruns irrespective of what trigger it is. IPC timeouts are fatal and need a reboot. This would avoid it.

**Review by lyakh — APPROVED** [review]

File: src/audio/pipeline/pipeline-schedule.c [hunk 1]
```diff
@@ -48,6 +48,25 @@ static enum task_state pipeline_task(void *arg)
...
+		 * this pipeline task gets a chance to run. But the host is still waiting for a
+		 * trigger IPC response. So, send an error response to prevent it from getting
+		 * timed out. No point triggering the pipeline in this case. It will be stopped
+		 * anyway by the host.
+		 */
+		if (p->trigger.cmd >= 0) {
+			struct sof_ipc_reply reply = {
+				.hdr.cmd = SOF_IPC_GLB_REPLY,
+				.hdr.size = sizeof(reply),
+				.error = -EINVAL,
```
Comment (plbossart) [inline src/audio/pipeline/pipeline-schedule.c:62]: nit-pick: is the error code relevant here? -EIO or -EPIPE could better describe the problem, no?

Comment (ranj063) [reply src/audio/pipeline/pipeline-schedule.c]: ack. fixed!

**Review by plbossart — APPROVED** [review]

**Review by lgirdwood — APPROVED** [review]

**Comment by lgirdwood** [comment]
Some untested DUTs, but CI tested all platforms.

**Pull request merged**